trigger/wsclient/trigger.go

```diff
@@ -42,28 +52,94 @@ func (*Factory) New(config *trigger.Config) (trigger.Trigger, error) {
 	}
-
 	return &Trigger{settings: s, config: config}, nil
 }
 
 // Initialize initializes the trigger
 func (t *Trigger) Initialize(ctx trigger.InitContext) error {
 	t.logger = ctx.Logger()
-	urlSetting := t.config.Settings["url"]
-	if urlSetting == nil || urlSetting.(string) == "" {
-		return fmt.Errorf("server url not provided")
+	headers := t.settings.Headers
+	queryParams := t.settings.QueryParams
+	urlstring := t.settings.URL
+	// populate headers
+	header := make(http.Header)
+	if len(headers) > 0 {
+		for key, val := range headers {
+			header[key] = []string{val}
+		}
+	}
+	// populate queryparam
+	if len(queryParams) > 0 {
+		qp := url.Values{}
+		for key, val := range queryParams {
+			qp.Add(key, val)
+		}
+		urlstring = urlstring + "?" + qp.Encode()
+	}
+	var isWSS bool
+	if urlstring != "" {
+		isWSS = strings.HasPrefix(urlstring, "wss")
+	}
+	var dialer websocket.Dialer
+	if isWSS {
+		tlsconfig := &tls.Config{}
+		allowInsecure := t.settings.AllowInsecure
+		if allowInsecure {
+			tlsconfig.InsecureSkipVerify = true
+		} else {
+			caCertValue := t.settings.CaCert
+			var cacertObj map[string]interface{}
+			if caCertValue != "" {
+				err := json.Unmarshal([]byte(caCertValue), &cacertObj)
+				if err != nil { // filepath configured
+					certPool, err := getCerts(caCertValue)
+					if err != nil {
+						t.logger.Infof("Error while loading client trust store - %v", err)
+						return err
+					}
+					tlsconfig.RootCAs = certPool
+				} else { // file content configured
+					rootCAbytes, err := decodeCerts(caCertValue, t.logger)
+					if err != nil {
+						t.logger.Error(err)
+						return err
+					}
+					certPool := x509.NewCertPool()
+					certPool.AppendCertsFromPEM(rootCAbytes)
+					tlsconfig.RootCAs = certPool
+				}
+			}
+		}
+		dialer = websocket.Dialer{TLSClientConfig: tlsconfig}
+	} else {
+		dialer = *websocket.DefaultDialer
 	}
 
-	url := urlSetting.(string)
-	t.logger.Infof("dialing websocket endpoint[%s]...", url)
-	conn, _, err := websocket.DefaultDialer.Dial(url, nil)
+	t.logger.Debugf("dialing websocket endpoint[%s] with headers[%s]...", urlstring, header)
+	conn, res, err := dialer.Dial(urlstring, header)
 	if err != nil {
-		return fmt.Errorf("error while connecting to websocket endpoint[%s] - %s", url, err)
+		if res != nil {
+			defer res.Body.Close()
+			body, err1 := ioutil.ReadAll(res.Body)
+			if err1 != nil{
+				ctx.Logger().Infof("res code is %v error while reading response payload is %s ", res.StatusCode,  err1)
+			}
+			t.logger.Infof("res code is %v payload is %s , err is %s", res.StatusCode, string(body), err)
+		}
+		return fmt.Errorf("error while connecting to websocket endpoint[%s] - %s", urlstring, err)
 	}
+
 	t.wsconn = conn
 	go func() {
+		defer func(){
+			err := conn.WriteMessage(websocket.CloseMessage, []byte("Sending close message while getting out of reading connection loop"))
+			if err != nil{
+				t.logger.Debugf("Received err [%s] while writing close message", err)
```

Comment:
Maybe the error level is properly here.

Reply:
converted to warn level as its not critical to be logged as error. done.